codec/src/main/java/io/netty/handler/codec/compression/BrotliEncoder.java

```diff
@@ -69,23 +140,141 @@ protected void encode(ChannelHandlerContext ctx, ByteBuf msg, ByteBuf out) throw
+        private final BrotliEncoderChannel brotliEncoderChannel;
+        private final ChannelHandlerContext ctx;
+        private boolean isClosed;
+
+        private Writer(Encoder.Parameters parameters, ChannelHandlerContext ctx) throws IOException {
+            brotliEncoderChannel = new BrotliEncoderChannel(this, parameters);
+            this.ctx = ctx;
+        }
+
+        private void encode(ByteBuf msg, boolean preferDirect) throws Exception {
+            try {
+                allocate(preferDirect);
+
+                // Compress data and flush it into Buffer.
+                //
+                // As soon as we call flush, Encoder will be triggered to write encoded
+                // data into WritableByteChannel.
+                //
+                // A race condition will not arise because one flush call to encoder will result
+                // in only 1 call at `write(ByteBuffer)`.
+                brotliEncoderChannel.write(msg.nioBuffer());
+                FLUSH_METHOD.invoke(brotliEncoderChannel);
+
+            } catch (Exception e) {
+                ReferenceCountUtil.release(msg);
+                throw e;
+            }
+        }
+
+        private void allocate(boolean preferDirect) {
             if (preferDirect) {
-                out = ctx.alloc().ioBuffer();
+                writableBuffer = ctx.alloc().ioBuffer();
             } else {
-                out = ctx.alloc().buffer();
+                writableBuffer = ctx.alloc().buffer();
+            }
+        }
+
+        @Override
+        public int write(ByteBuffer src) throws IOException {
+            if (!isOpen()) {
+                throw new ClosedChannelException();
+            }
+
+            return writableBuffer.writeBytes(src).readableBytes();
+        }
+
+        @Override
+        public boolean isOpen() {
+            return !isClosed;
+        }
+
+        @Override
+        public void close() {
+            final ChannelPromise promise = ctx.newPromise();
+
+            ctx.executor().execute(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        finish(promise);
+                    } catch (IOException ex) {
+                        promise.setFailure(new IllegalStateException("Failed to finish encoding", ex));
+                    }
+                }
+            });
+        }
+
+        public void finish(final ChannelPromise promise) throws IOException {
+            if (!isClosed) {
+                // Allocate a buffer and write last pending data.
+                allocate(true);
+                brotliEncoderChannel.close();
+
+                ctx.writeAndFlush(writableBuffer, promise).addListener(new ChannelFutureListener() {
+                    @Override
+                    public void operationComplete(ChannelFuture future) {
+                        if (future.isSuccess()) {
+                            isClosed = true;
```

Comment:
Why do we delay setting isClosed until after the write, and then only if the write succeeds?

Reply:
Changed the approach. PTAL. :)